pallets/proofs-dealer/src/utils.rs

```diff
@@ -49,22 +88,134 @@ where
         Self::enqueue_challenge(key)
     }
 
-    // TODO: Document and add proper parameters.
-    // TODO: Remove unused variable allow attribute.
-    #[allow(unused_variables)]
-    pub fn do_submit_proof(submitter: &ProviderFor<T>, proof: &CompactProof) -> DispatchResult {
+    /// Submit proof.
+    ///
+    /// For a given `submitter`, verify the `proof` submitted. The proof is verified by checking
+    /// the forest proof and each key proof.
+    /// Relies on the `ProvidersPallet` to get the root for the submitter, the last block for which
+    /// the submitter submitted a proof and the stake for the submitter. With that information, it
+    /// computes the next block for which the submitter should be submitting a proof. It then gets
+    /// the seed for that block and generates the challenges from the seed. It also checks if there
+    /// has been a Checkpoint Challenge block in between the last block proven and the current block.
+    /// If there has been, the Provider should have included proofs for the challenges in that block.
+    /// It then verifies the forest proof and each key proof, using the `ForestVerifier` and `KeyVerifier`.
+    pub fn do_submit_proof(submitter: &ProviderFor<T>, proof: &Proof<T>) -> DispatchResult {
+        let forest_proof = &proof.forest_proof;
+        let key_proofs = &proof.key_proofs;
+
         // Check if submitter is a registered Provider.
         ensure!(
             ProvidersPalletFor::<T>::is_provider(submitter.clone()),
             Error::<T>::NotProvider
         );
 
-        // TODO: Check that the proof corresponds to a correct challenge block.
+        // Check that key_proofs is not empty.
+        ensure!(!key_proofs.is_empty(), Error::<T>::EmptyKeyProofs);
 
-        // TODO: Verify proof.
+        // The check for whether forest_proof and each key_proof is not empty is handled by the corresponding
+        // verifiers for each. We do not preemptively check for this here, since the `CommitmentVerifier::Proof`
+        // type is not required to have an `is_empty` method.
```

Comment:
Is this a lingering comment?

Reply:
No, this is a comment for why I don't start by checking if the proofs are empty. Because if I could, I would do it here, to avoid all the rest of the comptuation.